packages/funding-rate-proposer/src/proposer.js

```diff
@@ -0,0 +1,360 @@
+            : "Failed to propose funding rate🚨";
+        this.logger.error({
+          at: "PerpetualProposer#updateFundingRate",
+          message,
+          fundingRateIdentifier,
+          error
+        });
+        return;
+      }
+    } else {
+      this.logger.debug({
+        at: "PerpetualProposer#updateFundingRate",
+        message: "Skipping proposal because current rate is within allowed margin of error",
+        fundingRateIdentifier,
+        proposedRate: offchainFundingRate,
+        currentRate: onchainFundingRate,
+        allowedError: this.fundingRateErrorPercent
+      });
+    }
+  }
+  // Sets allowances for all collateral currencies used live perpetual contracts.
+  async _setAllowances() {
+    // Get new approval receipts or null if approval was unneccessary.
+    const newApprovals = await Promise.map(Object.keys(this.contractCache), contractAddress => {
+      // The Perpetual requires approval to transfer the contract's collateral currency in order to post a bond.
+      // We'll set this once to the max value and top up whenever the bot's allowance drops below MAX_INT / 2.
+      return setAllowance(
+        this.web3,
+        this.gasEstimator,
+        this.account,
+        contractAddress,
+        this.contractCache[contractAddress].collateralAddress
+      );
+    });
+    newApprovals.forEach(receipt => {
+      if (receipt) {
+        this.logger.info({
+          at: "PerpetualProposer",
+          message: "Approved Perpetual contract to transfer unlimited collateral tokens 💰",
+          perpetual: receipt.spenderAddress,
+          currency: receipt.currencyAddress,
+          collateralApprovalTx: receipt.tx.transactionHash
+        });
+      }
+    });
+  }
+
+  async _cacheAndUpdatePriceFeeds() {
+    for (let contractAddress of Object.keys(this.contractCache)) {
+      const fundingRateIdentifier = this.hexToUtf8(
+        this.contractCache[contractAddress].state.currentFundingRateData.identifier
+      );
+      let priceFeed = this.priceFeedCache[fundingRateIdentifier];
+      if (!priceFeed) {
+        this.logger.debug({
+          at: "PerpetualProposer",
+          message: "Caching new pricefeed for identifier",
+          commonPriceFeedConfig: this.commonPriceFeedConfig,
+          fundingRateIdentifier
+        });
+
+        // Create a new pricefeed for this identifier. We might consider caching these price requests
+        // for re-use if any requests use the same identifier.
+        priceFeed = await createReferencePriceFeedForFinancialContract(
+          this.logger,
+          this.web3,
+          new Networker(this.logger),
+          () => Math.round(new Date().getTime() / 1000),
+          null, // No EMP Address needed since we're passing identifier explicitly
+          this.commonPriceFeedConfig,
+          fundingRateIdentifier
+        );
+        this.priceFeedCache[fundingRateIdentifier] = priceFeed;
```

Comment:
What if this comes back null (because we don't know how to construct this pf)? Should we send an error log? Do we just expect the bot to crash? It seems like we should probably just fire off an error log and keep going?

Reply:
Yeah I think we should skip setting it in the cache during this method, and then emit an error log in the `updateFUndingRates()` call

Reply:
+1, that sounds like the right way to do it

Reply:
You don't even need to skip setting it here since it'll set it to null, and the other method should be able to recognize that null.

Reply:
- I added an `if(pricefeed)` before the subsequent `await pricefeed.update()`
- I catch `!pricefeed` in the `updateFundingRates()` method and emit error log (and exit early) if no pricefeed.

Reply:
Bump.

Reply:
Ignore this^. LGTM